src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +113,258 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+                }
+            }
+        }
+
+        /// <summary>
+        /// Generates a cache key for Transitive Originas cache
+        /// </summary>
+        /// <param name="transitivePackage"></param>
+        /// <returns>A string with given key</returns>
+        /// <seealso cref="GetCachedTransitiveOrigin(PackageIdentity)"/>
+        /// <seealso cref="SetCachedTransitiveOrigin(PackageIdentity, TransitiveEntry)"/>
+        internal string GetTransitiveCacheKey(PackageIdentity transitivePackage)
+        {
+            return _projectUniqueName + "/" + transitivePackage.Id.ToLowerInvariant() + "." + transitivePackage.Version.ToNormalizedString();
+        }
+
+        /// <summary>
+        /// Obtains cached entry for a given transitive package
+        /// </summary>
+        /// <param name="transitivePackage">Identity of transitive package</param>
+        /// <returns>A <see cref="TransitiveEntry"/> object, or <c>null</c> if not found</returns>
+        /// <seealso cref="ClearCachedTransitiveOrigins"/>
+        /// <seealso cref="SetCachedTransitiveOrigin(PackageIdentity, TransitiveEntry)"/>
+        internal TransitiveEntry GetCachedTransitiveOrigin(PackageIdentity transitivePackage)
+        {
+            string key = GetTransitiveCacheKey(transitivePackage);
+
+            if (TransitiveOriginsCache.ContainsKey(key))
+            {
+                return TransitiveOriginsCache[key];
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Replaces cached entry for a given transitive package with a <see cref="TransitiveEntry"/>
+        /// </summary>
+        /// <param name="transitivePackage">Identity of transitive package</param>
+        /// <param name="origins">Packages identified as package origins</param>
+        /// <seealso cref="ClearCachedTransitiveOrigins"/>
+        /// <seealso cref="GetCachedTransitiveOrigin(PackageIdentity)"/>
+        internal void SetCachedTransitiveOrigin(PackageIdentity transitivePackage, TransitiveEntry origins)
+        {
+            string key = GetTransitiveCacheKey(transitivePackage);
+            TransitiveOriginsCache[key] = origins;
+        }
+
+        /// <summary>
+        /// Clears Transitive Origins cache
+        /// </summary>
+        /// <seealso cref="GetCachedTransitiveOrigin(PackageIdentity)"/>
+        /// <seealso cref="SetCachedTransitiveOrigin(PackageIdentity, TransitiveEntry)"/>
+        internal void ClearCachedTransitiveOrigins()
+        {
+            TransitiveOriginsCache.Clear();
+            IsTransitiveComputationNeeded = true;
+        }
+
+        internal static TransitivePackageReference MergeTransitiveOrigin(PackageReference currentPackage, TransitiveEntry transitiveEntry)
+        {
+            var transitiveOrigins = new SortedSet<PackageReference>(PackageReferenceMergeComparer);
+
+            if (transitiveEntry != null)
+            {
+                foreach (var key in transitiveEntry.Keys)
```

Comment:
Just my opinion, not blocking, linq query would be more concise.

Reply:
Fixed.